### PR TITLE
Add reshape+batching and dynamic batching support for string I/O

### PR DESCRIPTION
**Pull request by krishung5**
Test added here: https://github.com/triton-inference-server/server/pull/4574

### CODE REVIEW

File: src/libtorch.cc [hunk 1]
```diff
@@ -1532,9 +1527,7 @@ SetStringInputTensor(
                   std::to_string(element_idx + 1) + " for inference input '" +
                   name + "', expecting " + std::to_string(request_element_cnt))
                   .c_str()));
-      FillStringTensor(
-          input_list, tensor_offset + element_idx,
```
Comment (GuanLuo) [inline src/libtorch.cc:1536]: `tensor_offset` is no longer needed? If it becomes unused parameter, then you can remove it from the function signature and propagate the change up on the call stack.

File: src/libtorch.cc [hunk 2]
```diff
@@ -1832,6 +1813,7 @@ ModelInstanceState::ReadOutputTensors(
 
     } else if (output_tensors[op_index].isList()) {
       // Custom handling for string/bytes tensor...
+      const int max_batch_size = model_state_->MaxBatchSize();
```
Comment (GuanLuo) [inline src/libtorch.cc:1816]: You can store `support_batching` as member variable, then you don't need to compute the value again and again

File: README.md [hunk 3]
```diff
@@ -214,5 +214,3 @@ a List of Strings as input(s) / produces a List of String as output(s). For thes
 Triton allows users to pass String input(s)/recieve String output(s) using the String
 datatype. As a limitation of using List instead of Tensor for String I/O, only for
 1-dimensional input(s)/output(s) are supported for I/O of String type.
-Batching is not allowed for PyTorch models with String I/O. For these models,
```
Comment (CoderHam) [inline README.md:217]: We still cannot allow 2D string I/O in the pytorch framework. Modify the docs to highlight the same

File: src/libtorch.cc (hunk 2 above)
Comment (CoderHam) [reply src/libtorch.cc:1816]: +1

**Review by CoderHam — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (krishung5) [reply README.md:217]: I think this should be addressed already in the previous line where we mention the framework only supports 1D string I/O?

File: src/libtorch.cc (hunk 1 above)
Comment (krishung5) [reply src/libtorch.cc:1536]: Thanks for the catch. Removed the unused variable.

File: src/libtorch.cc (hunk 2 above)
Comment (krishung5) [reply src/libtorch.cc]: Done

File: src/libtorch.cc [hunk 4]
```diff
@@ -1532,9 +1528,7 @@ SetStringInputTensor(
                   std::to_string(element_idx + 1) + " for inference input '" +
                   name + "', expecting " + std::to_string(request_element_cnt))
                   .c_str()));
-      FillStringTensor(
-          input_list, tensor_offset + element_idx,
-          request_element_cnt - element_idx);
+      FillStringTensor(input_list, request_element_cnt - element_idx);
```
Comment (GuanLuo) [inline src/libtorch.cc:1531]: I don't think you need to fill the rest of string elements in this case, when `element_idx >= request_element_cnt` evaluates to true, there is `element_idx+1` elements in `input_list`

File: src/libtorch.cc [hunk 5]
```diff
@@ -1576,9 +1568,7 @@ SetStringInputTensor(
                           " strings for inference input '" + name + "', got " +
                           std::to_string(element_idx))
                           .c_str()));
-    FillStringTensor(
-        input_list, tensor_offset + element_idx,
-        request_element_cnt - element_idx);
+    FillStringTensor(input_list, request_element_cnt - element_idx);
```
Comment (GuanLuo) [inline src/libtorch.cc:1571]: Same here, only need to fill if `element_idx < request_element_cnt`, so add an additional check right before `FillStringTensor`

File: src/libtorch.cc (hunk 4 above)
Comment (krishung5) [reply src/libtorch.cc]: Thanks for the catch. Removed this line.

File: src/libtorch.cc (hunk 5 above)
Comment (krishung5) [reply src/libtorch.cc]: Added a check for this

**Comment by krishung5** [comment]
CI passes: https://gitlab-master.nvidia.com/dl/dgx/tritonserver/-/pipelines/5274652 please review, thanks!

**Review by CoderHam — APPROVED** [review]

**Review by GuanLuo — APPROVED** [review]

**Pull request merged**